FWCore/Framework/interface/Principal.h

```diff
@@ -80,7 +80,9 @@ namespace edm {
 
     void adjustIndexesAfterProductRegistryAddition();
 
+    void fillPrincipal(DelayedReader* reader);
```

Comment:
Are these different `fillPrincipal` needed for different inheriting types? If so, then having them in the base class feels wrong.

Reply:
The first function is the common part that is the same for all four types of Principals. There is no problem with that one.

Yes. The second and third functions are needed for different inheriting types. The reason is that there is not a unique ProcessHistory associated with a ProcessBlock. This is a real and fundamental difference.

I agree this "feels" wrong. I was not completely happy with what I did here. I can see three choices.

1. Leave it as it is. Just live with it. It is probably the simplest solution.

2. I could combine the two functions into one function. When called from ProcessBlockPrincipal I could pass in dummy arguments and with an if-else block or two I could probably make it work. To be honest that still kind of feels "wrong", but it would be less obvious and hidden inside the function (not in the interface).

3. I could add a new class to the inheritance hierarchy. RunPrincipal inherits from PrincipalUsingProcessHistory which inherits from Principal. Same for EventPrincipal and LuminosityBlockPrincipal. Then move the content of one of the functions up into PrincipalUsingProcessHistory.  In addition, move the functionality in the other function up into ProcessBlockPrincipal since that is the only type not using the ProcessHistory. It would takes some time to work out the details, but it should work.

Which would you prefer? Or do you have a better idea?

Reply:
I responded to the comment on Principal.h related to the fillPrincipal functions. I just wanted to highlight this discussion because I could use some more feedback on what you want me to do. I am leaning towards leaving the PR as it is, but I could add this extra class in the inheritance hierarchy or do something else.

It feels a little bit like I pounded a square block in a round whole here and in one or two other places, because there are some differences between ProcessBlocks and Runs/Lumis/Events. But the alternatives introduce so much complexity that I think just living with this little imperfection is best. We could change this later if we find it is actually causing problems.

Reply:
Let's leave it as it is now, and see later if we get a better idea for the code organization.